src/main.nr

```diff
@@ -58,35 +69,74 @@ contract CoinToss {
 
     // Constructs the contract.
     #[aztec(private)]
-    fn constructor(divinity: AztecAddress, private_oracle: AztecAddress, house: AztecAddress, bet_amount: Field) {
-        let mut config_note = ConfigNote::new(divinity, private_oracle, house, bet_amount);
+    fn constructor(divinity: AztecAddress, private_oracle: AztecAddress, house: AztecAddress, token: AztecAddress, bet_amount: Field) {
+        let mut config_note = ConfigNote::new(divinity, private_oracle, house, token, bet_amount);
         storage.config.initialize(&mut config_note, Option::none(), false);
     }
 
+    // House shares off-chain with the user the preimage for an escrow note for the bet_amount and an authwit to consume it
     #[aztec(private)]
-    fn create_bet(bet: bool) {
+    fn create_bet(bet: bool, userTransferNonce: Field, houseEscrowRandomness: Field, houseSettleEscrowNonce: Field) {
         let sender = AztecAddress::new(context.msg_sender());
-        let house = storage.config.get_note().house;
-        let address_this = AztecAddress::new(context.this_address()).address;
+        let config_note = storage.config.get_note();
+        let house = config_note.house;
+        let token = config_note.token;
+        let bet_amount = config_note.bet_amount;
+        let address_this = AztecAddress::new(context.this_address());
+
+        let token_contract = TokenPrivateContextInterface::at(token.address);
+
+        // We transfer private tokens from the user to the coin toss 
+        let _ = token_contract.transfer(
+            &mut context,
+            FromTransferStruct { address: sender.address },
+            ToTransferStruct { address: address_this.address },
+            bet_amount,
+            userTransferNonce
+        );
+
+        // We consume the escrow from the house
+        let __ = token_contract.settle_escrow(
+            &mut context,
+            OwnerSettleEscrowStruct { address: house.address },
+            RecipientSettleEscrowStruct { address: address_this.address }, 
+            houseEscrowRandomness,
+            houseSettleEscrowNonce
+        );
```

Comment:
I didn't think it was necessary just because of the transient balance. We know exactly how many tokens the user sent and if the escrow from the house didn't match it, it would revert because of not enough balance

Reply:
We could add it to have a more expressive revert maybe?

Reply:
@0xShaito I’ll add tests for every sad paths and check if errors are informative enough (sad paths aren’t really covered rn, I’ll try to complete our tested paths to be ecxhaustive, I’m just afraid of the test running time)- wdyt?

Reply:
I'm not too worried about expressive reverts at this stage (they are important though). As for sad paths, agree on covering them, but I'd focus on having the happy path working first.

Still, the scenarios I was worried about are:
- If later down the road there's a way for the cointoss contract to actually hold funds of its own, the `settle_escrow` could send not enough money to the contract, and end up siphoning its own funds.
- If the house accidentally set up an escrow for _more_ funds, we should have an extra check to prevent it from getting those funds lost/locked in the cointoss contract.